Implement an OH2 EnOcean binding (Closes #2261)

fruggy83 opened the pull request:
Closes #2261

Signed-off-by: Daniel Weber <uni@fruggy.de>
also-by: Aitor Iturrioz <riturrioz@gmail.com>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

[enocean] Initial contribution

This PR provides a new version of the EnOcean binding for OH2.

Motivation:
This PR is based on the unofficial [OpenOcean Binding](https://github.com/fruggy83/openocean) for OH2. I implemented this binding to use the new features of OH2 like auto discovery and to be able to send telegrams to actors by using device specific EEPs rather than simple rocker switch telegrams.
To make this binding accessable for a wider range of users I implemented this PR.

Goal:
The goal is to implement a fully functional and easy expandable binding to support a wide range of EnOcean enabled devices without using any third party lib or introducing new dependencies.

Details:
This binding implements a full functional EnOcean stack for receiving and sending telegrams. It is developed on an EnOcean USB300 and runs on an EnOceanPi too. The following features are already implemented:
- Serial and TCP bridge
- Auto discovery of EnOcean USB300 serial bridge
- Auto discovery of many different device types like rollershutters, smart plugs, temperature sensors and many more
- Profiles for linking a rocker switch to a player item and a switch item to a virtual rocker switch to emit rocker switch telegrams

During review please have in mind that I am not a JAVA expert and I used this project to get more into JAVA too. So please review especially the threading parts carefully. Do not hesitate to ask my any question regarding this code

TODOs:
- Source code documentation
- The binding documentation itself is not up to date right now (readme.md)
- Translations
- Tests

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+ * Hoppe SecuSignal window handles
+ * Rocker switches (NodOn, Eltako FT55 etc)
+
+However because of the standardized EnOcean protocol it is more important which EEP this binding supports: F6-02, F6-10, D5-00, A5-10, A5-38 (switching and dimming), A5-37 (use generic things). Hence if your device supports one of these EEPs the chances are good that your device is also supported by this binding.
```
You should provide a table with the exact thing type ids and their description.

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
````diff
@@ -0,0 +1,128 @@
+For an inbound transformation (Enocean message => Openhab state) you get the channel id and the Enocean data seperated by a pipe.
+
+```
+ChannelId|EnoceanData(Hex)=OpenhabState|Value
````
This isn't clear to me.  Where is this information entered?

fruggy83 replied:
I am using the transform functionality of openHAB in this case. So you have to enter these information in files and put them into the transform folder of openHAB. Updated the doc regarding this point.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Channels
+
+The channels of a thing are determined automatically based on the choosen EEP. The following channels are supported: (Light) Switch, Dimmer, Rollershutter, Temperature, Handle state and many more.
```
Here as well: Please provide a table of all channes here and list, which thing type has which channels.

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
````diff
@@ -0,0 +1,128 @@
+```
+
+or link them directly to your items. The following items are supported by profiles: switch, dimmer, player.
+
````
Please add a "Full Example" section (see other bindings)

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+
+Bridge
+
+ * Serial port: The serial port to which the EnOcean gateway is connected to
```
You also define a tcpBridge, this should be mentioned here as well.

fruggy83 replied:
fixed, tcpBridge could be removed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,46 @@
+ org.openhab.binding.enocean.profiles
+Import-Package: 
+ com.google.common.collect,
+ gnu.io,
```
We by now have an ESH serial API (see [here](https://github.com/eclipse/smarthome/tree/master/bundles/io/org.eclipse.smarthome.io.transport.serial/src/main/java/org/eclipse/smarthome/io/transport/serial)), which should be used instead of gnu.io directly (it is a facade, which in the background still uses gnu.io for openHAB).

fruggy83 replied:
I switched over to the ESH serial API (nice work done here). Using this API I could remove the tcpBridge as this API supports rfc2217. However as long as the NRJavaSerial Bug is not fixed on windows OS, I have to stick with gnu.io. I commented these lines with a TODO. On linux machines the ESH serial API is used. However I get an exception when I try to use rfc2217 (gnu.io.serialport missing). Maybe I have done something wrong with the integration of this API.

kaikreuzer replied:
See my comment below - nrjavaserial should also work on Windows, so no need for gnu.io import anymore.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,46 @@
+ org.openhab.binding.enocean.handler,
+ org.openhab.binding.enocean.profiles
+Import-Package: 
+ com.google.common.collect,
```
We try to get rid of Guava, so please try to do without this dependency.

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,46 @@
+ org.eclipse.smarthome.core.transform,
+ org.eclipse.smarthome.core.transform.actions,
+ org.eclipse.smarthome.core.types,
+ org.junit;version="4.0.0";resolution:=optional,
```
this should not be needed

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,46 @@
+ org.openhab.binding.enocean.handler,
+ org.openhab.binding.enocean.profiles,
+ org.osgi.framework,
+ org.osgi.service.component.annotations;resolution:=optional,
```
this can be removed

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/tcpbridge.xml:
```diff
@@ -0,0 +1,51 @@
+			<channel typeId="setBaseId" id="setBaseId" />
+		</channels>
+
+		<properties>
```
remove those - properties should only be listed with static values here.

fruggy83 replied:
removed all properties

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/tcpbridge.xml:
```diff
@@ -0,0 +1,51 @@
+				<description>Host where ESP3 device is connected.</description>
+				<context>network-address</context>
+			</parameter>
+			<parameter name="port" type="integer" required="true">
```
not required as you have a default value

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/serialbridge.xml:
```diff
@@ -0,0 +1,47 @@
+			<channel typeId="setBaseId" id="setBaseId" />
+		</channels>
+
+		<properties>
```
remove those - properties should only be listed with static values here.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/tcpbridge.xml:
```diff
@@ -0,0 +1,51 @@
+
+	<!-- EnOcean Bridge -->
+	<bridge-type id="tcpbridge">
+		<label>EnOcean TCP Bridge</label>
```
In the readme, you call those an "EnOcean Gateway" - what is the official naming?

fruggy83 replied:
Changed to official naming Gateway (messed up with openHAB wording)

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/serialbridge.xml:
```diff
@@ -0,0 +1,47 @@
+
+	<!-- EnOcean Bridge -->
+	<bridge-type id="serialbridge">
+		<label>EnOcean Serial Bridge</label>
```
In the readme, you call those an "EnOcean Gateway" - what is the official naming?

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,268 @@
+	</channel-type>
+
+	<!-- virtual rocker switch channel -->
+	<channel-type id="virtualRockerswitch">
```
Doesn't this duplicate the trigger channel? In general no "virtual" channels should be provided by a binding.

fruggy83 replied:
A VirtualRockerSwitch is used to send rocker switch messages (uses switch channels acts like a virtual rocker). A RockerSwitch receives messages (uses trigger channels, represents a physical rocker). Had a long discussion about seperating these concepts into two thingtypes to make it more simple. However I am open for any new suggestions.

kaikreuzer replied:
Hm, ok, I think I understand what it is technically doing, but let me try to come from the use case side: Having a virtual rocker by itself does not bring any value, it will just send telegrams that go to nirvana. So a user will actually always control something with it - so effectively, the state of the linked item reflects the state of the device that is switched (and not that of the virtualRocker, which actually does not have any state). So if we are switching other devices, my question is whether we cannot control those other devices through their own channels already (and if not, if such "control channels" could/should be added to them?

> Had a long discussion about seperating these concepts 

Is this somewhere available on Github? If so, please point me to it and I might ask less stupid questions in future ;-)

fruggy83 replied:
This rocker switch simulation (VirtualRockerSwitch) is just meant for old EnOcean devices which do not support the more advanced and newer EEPs like CentralCommand. 

I really like the idea to hide this technical aspect by using the already existing things and configure them to use rocker switch simulation instead of their advanced EEPs. However as I defined the thing-types according to their EEP they use, I have to introduce a new thing-type *OldEnOceanDeviceControlledByRockerSwitchMessages*. This thing-type would have at least a switching, rollershutter and dimmer channel. Good point.
However to be still able to emit pure rocker switch messages I would like to keep the VirtualRockerSwitch and its switch profile as an unlisted thing-type. By the way are not the profiles intended for such a use case (do the transfer logic from items to things and vice versa)?

You can find a part of the discussion [here](https://github.com/fruggy83/openocean/issues/6).

fruggy83 replied:
I changed the VirtualRockerSwitch thing label to Classic Device, as this thing is essentially intended for such old devices, which just support rocker switch messages. I did not change the channels and kept the virtualrockerswitch ones. However I added a new profile for linking these channels to a rollershutter item (reacts only to up/down command). In this way the advanced users can still use these channels in their rules (some users from the community are using these things already in this way) and normal user are able to simple link these channels to switch/rollershutter items. A dimmer profile can be discussed later ;).
ClassicDevices and RockerSwitches should be used together
RockerSwitch => system profile => Switch Item => switch profile => ClassicDevice

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,268 @@
+		</state>
+	</channel-type>
+
+	<channel-type id="setBaseId" advanced="true">
```
Sounds like a candidate for https://github.com/eclipse/smarthome/issues/3484 instead of being a channel.

fruggy83 replied:
Yes you are absolutely right. However as you can only change this BaseId 10 times, I wanted to make this fool proof and setup an advanced channel for that. In this way the user explicitly has to link and interact with it.

kaikreuzer replied:
ok, sounds like the safest solution then.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/CentralCommand.xml:
```diff
@@ -0,0 +1,54 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="centralCommand">
```
This does not sound as if it represents any kind of real device? Isn't this rather a channel of devices that also implement other EEPs at the same time?

fruggy83 replied:
I have choosen the EEP family name from the enocean documentation for the thing-type ids. A central command device is a device which sends (central ;)) commands to different kinds of actuators like simple light switches, dimmers, fans or thermostats. As suggested I could also use the EEP number for the thing-type id.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/CentralCommand.xml:
```diff
@@ -0,0 +1,54 @@
+		<label>A5-38 - Central command</label>
+		<description>Send a command to an actor by using A5-38 EEP</description>
+
+		<properties>
```
remove

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/CentralCommand.xml:
```diff
@@ -0,0 +1,54 @@
+			<bridge-type-ref id="tcpbridge" />
+		</supported-bridge-type-refs>
+
+		<label>A5-38 - Central command</label>
```
The labels should be the name of the physical device that has been discovered (or selected to be added).
The names should be user-friendly, i.e. "A5-38 -" should not be here; if relevant to tell two different devices apart, you might add it in brackets to the end, but usually it should suffice in the description.

fruggy83 replied:
As the EEP number is so commonly used in device specs, I wanted to place it directly in the name/label.

kaikreuzer replied:
How about at least changing the order, e.g. to "Central Command (A5-38)"? Btw, we usually use capitalisation in Labels, so you might want to adapt them.

fruggy83 replied:
Removed EEP reference from label and added it to the description. Labels use capitalisation now.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/MeasurementSwitch.xml:
```diff
@@ -0,0 +1,51 @@
+			<bridge-type-ref id="tcpbridge" />
+		</supported-bridge-type-refs>
+
+		<label>D2-01 - Energy measurement switch</label>
```
Usually, such a device would be called "EnOcean Smart Plug" or something.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/ContactAndSwitch.xml:
```diff
@@ -0,0 +1,31 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="contactSwitch">
```
Wouldn't it make sense to choose the thing type id to be the eep, i.e. here "D5-00"?

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/ContactAndSwitch.xml:
```diff
@@ -0,0 +1,31 @@
+
+		<label>D5-00 - Single input contact</label>
+		<description>Single input contact which sends EEP D5-00-01</description>
+
```
Why are there no channels defined for this thing type? Same question for the others.

fruggy83 replied:
A thing type can handle EEPs of a certain EEP family. However the EEPs (although belonging to the same family) provide different functionality. For example the EEP family RoomOperatingPanel A5-10 has profiles which can set a fan stage, set an occupancy switch or send temperature or humidity sensor data. As I did not want to create different thing types of all of these combinations, I create the channels dynamically upon the current choosen EEP. In this way I only need one thing type => RoomOperatingPanel.

kaikreuzer replied:
ok, thanks.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/HumidityTemperatureSensor.xml:
```diff
@@ -0,0 +1,32 @@
+			<bridge-type-ref id="tcpbridge" />
+		</supported-bridge-type-refs>
+
+		<label>A5-04 - Temperature and humidity sensor</label>
```
Also better "EnOcean Temperature/Humidity Sensor"

fruggy83 replied:
fixed wording

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/VirtualRockerSwitch.xml:
```diff
@@ -0,0 +1,48 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- RockerSwitch -->
+	<thing-type id="virtualRockerSwitch">
```
There should be no virtual thing types.

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,268 @@
+	</channel-type>
+
+	<!-- Light Switch Channel -->
+	<channel-type id="lightSwitch">
```
Isn't there any numbering/naming convention in the EnOcean spec for the single values? I would expect the channel types here to directly refer to parts of the spec (and also have their IDs chosen that way).

fruggy83 replied:
fixed => set to generalSwitch

fruggy83 commented:
Hi @kaikreuzer,

thanks a lot for your valuable feedback. I think it is the best if I first update the readme.md and answer your questions before you dive deeper into the source code.

Best regards
Daniel

fruggy83 commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+# EnOcean Binding
+
+The EnOcean binding connects OpenHab to the EnOcean ecosystem.
```
Should be solved now

fruggy83 commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+
+The EnOcean binding connects OpenHab to the EnOcean ecosystem.
+
+The binding uses an EnOcean gateway to retrieve sensor data and control actuators. For _bidirectional_ actuators it is even possible to update the OpenHab item state if the actuator gets modified outside of OpenHab.
```
Should be solved now

fruggy83 commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Thing Configuration
+
+Things can and should by configured through PaperUI. Following the most important config parameters:
```
I updated the documentation regarding this point. Using PaperUI is particularly needed for the pairing process. However after pairing is done, you are able to configure your things in files.

Just a side note: I think that parameters with multiple options cannot be set in the thing file (see [#6146](https://github.com/eclipse/smarthome/issues/6146)). Implemented a workaround for this in my binding, but should be solved in ESH.

kaikreuzer replied:
Please correct "PaperUI" to "Paper UI".

> Using PaperUI is particularly needed for the pairing process.

Ok, understood for the devices that need pairing - but if I am not mistaken, most sensors (without security) do not require a pairing process, so you should formulate it in a way so that it is clear that it is only relevant for pairing; probably best to have a dedicated section on pairing and leave the "thing configuration" section independent of whether DSLs or UIs are used.

fruggy83 commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+Bridge
+
+ * Serial port: The serial port to which the EnOcean gateway is connected to
+ * Next device id: The device Id which should be taken for the next created actuator without an explicit device Id
```
I updated the doc, hope this makes things clear

fruggy83 commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,128 @@
+ 
+Things
+
+ * Thing Id: EnOcean Id of the device. 4 byte hex string (abcdef00).
```
fixed

fruggy83 commented:
Hi @kaikreuzer,

sorry for my late answer and push, I was very busy the last days.

I think I have fixed most of your points, some are still open to discuss. Thanks a lot for the hint regarding ESH serial API. I could drop the tcpBridge. However because of the NRJavaSerial bug I still need gnu.io on my windows machine.

Best regards
Daniel

kaikreuzer commented:
> However because of the NRJavaSerial bug I still need gnu.io on my windows machine.

The distro comes with JNA packaged, so nrjavaserial should work just fine on Windows as well.
If you are talking about the IDE, you can put the [JNA libs](https://github.com/java-native-access/jna#jna) directly in launch/home/lib (or whereever your `-Djava.library.path` parameter in your launch config points to).

fruggy83 commented:
Hello @kaikreuzer,

thanks again for your feedback. Especially regarding the parameter `-Djava.library.path`

Regarding the ThingId: It is not clear to me if I am allowed to use the EnOceanId as the ThingId during discovery or if I have to use/implement an algorithm which creates unique ids for me?
Furthermore if the ThingId should be just a unique random number, how can I detect if a new discovered device already exists? What is the intended use of DiscoveryServiceCallback.getExistingThing(thingUID), if I have to iterate over all things of my bridge and check their DeviceId?

I hope I find time in the next days to solve the mentioned issues.

kaikreuzer commented:
> It is not clear to me if I am allowed to use the EnOceanId as the ThingId during discovery

Yes, this is perfectly fine if it serves you as a means to make sure that you do not produce duplicate IDs.

>  how can I detect if a new discovered device already exists? 

You should define your "deviceID" property as the representationProperty (see [here](https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html) of the discovery result - the framework will then know that it is a duplicate and not add it to the inbox.

> What is the intended use of DiscoveryServiceCallback.getExistingThing(thingUID)

This is [documented here](https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html#compare-with-existing-results). It is really only meant for very specific bindings/requirements, where you can only discover parts of a thing at a time and you have to check, what information you might have found already and how to merge that into a consistent result. This should imho not be relevant for the EnOcean binding.

fruggy83 commented:
Hi @kaikreuzer,
thanks a lot for your explanations and hint regarding the representationProperty, I was not aware of. Sorry for asking such dumb questions. I definitely have to read more docs regarding ESH and not only openHAB!

kaikreuzer commented:
All fine, I know that it isn't easy to find the answers in the docs as it isn't really well-structured. So I'm here to help (and learn where we might need to improve the docs)!

xxorde commented:
Hi @fruggy83 did you made any progress in the meantime?

kaikreuzer commented:
I'd be interested, too 😎

fruggy83 commented:
Hi all,

I think I have fixed all currently open issues in the meantime. Binding runs stable in my "productive" environment since the last few weeks. However I just wanted to wait till eclipse/smarthome#6386 found its way into openhab, so I can remove the workarounds and update the readme accordingly.

Best regards
Daniel

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/profiles/EnOceanProfileFactory.java:
```diff
@@ -0,0 +1,116 @@
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
```
Please add some JavaDoc to all classes in this package!

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/profiles/EnOceanProfileFactory.java:
```diff
@@ -0,0 +1,116 @@
+            if (CoreItemFactory.PLAYER.equalsIgnoreCase(itemType)) {
+                return EnOceanProfileTypes.RockerSwitchToPlayPause;
+            } else if (CoreItemFactory.ROLLERSHUTTER.equalsIgnoreCase(itemType)) {
+                return EnOceanProfileTypes.RockerSwitchToRollershutter;
```
Does this actually work on a `SYSTEM_RAWROCKER`? The implementation seems to expect events such as DIR1 and DIR2, so this looks rather EnOcean specific. If it is NOT EnOcean-specific, such profiles should rather be moved to ESH.

fruggy83 commented:
Hi @kaikreuzer 

> The framework gets hold of them by the fact that you register services.

Thanks for the explanations, I will move them to an internal package, as I do not believe that it makes sense that other bundles make use of these classes.

> although this channel type [SYSTEM_RAWROCKER] will only emit PRESSED and RELEASED events

Please have a look at the `DefaultSystemChannelTypeProvider` class. SYSTEM_RAWROCKER emits CommonTriggerEvents.DIR1/2_PRESSED/RELEASED events and SYSTEM_RAWBUTTON emits only CommonTriggerEvents.PRESSED/RELEASED.

> So this would be the example

Yes I am referring to this example. However you could also imagine a light of a corridor. Such lights are switched ON/OFF normally by more than a rocker switch. In my case I can switch my corridor light by four rocker switches.

> So you have linked the "classicDevice" [...] to this item as well as a real physical rocker, is this correct?

Yes that is correct. If I own just an older enocean actuator which only reacts on rocker switch messages, I have to use a classicDevice thing to control this light from within openHAB. First part of the link. 

> why do you have the physical rocker linked to the item at all? Is it for being informed about "external" commands to the classic device in order to update the item state accordingly?

Yes that is correct. The second part is just needed for state syncing, when the light gets switched from outside of openHAB through one of the physical rocker switches. Old enocean actuators do not send a status changed message, so we have to listen to the physical rocker switches to sync the item state with the physical state. To keep it simple for the user, I also use the rockerSwitch things in this case. So when you switch the item in openHAB it sends an ON command, which is interpreted by my RockerSwitchFromOnOff and translated into an enocean rocker PRESSED and RELEASED message. When you press a phyiscal rocker, the rockerSwitch thing emits a CommonTriggerEvents.DIR1_PRESSED event, which is interpreted by the (system) RawRockerToOnOffProfile and translated into an sendCommand(ON). Therefore the PHYSICAL light receives an ON command from the pyhsical rocker switch and another ON command from openHAB through interaction of both profiles. As I said this is not a big problems for lights. However if an enocean rollershutter receives two UP commands it will stop moving.

Option (1) came also into my mind. However the corridor light use case needs a user defined number of rocker switch addresses. This should not be a problem for textual configuration now, but I am not sure if this can be accomplished in PaperUI? I am also not a big fan of option (2) but it is maybe the most flexible solution. Otherwise if we say that the corridor light is just an edge use case, I could advance the classicDevice by one additional rocker switch address and for all other rocker switches the user has to create (postUpdate) rules. So I would like to let you the decision which way to go :)

kaikreuzer commented:
> Please have a look at the DefaultSystemChannelTypeProvider class. SYSTEM_RAWROCKER emits CommonTriggerEvents.DIR1/2_PRESSED/RELEASED events and SYSTEM_RAWBUTTON emits only CommonTriggerEvents.PRESSED/RELEASED.

Oh, you're right, I better should have had a look at the code. I somehow mixed up the RAWROCKER wiht the RAWBUTTON and wasn't really anymore aware that we had the system rockers defined as well, sorry. So all fine! 😇 
But in this case, I would indeed very strongly suggest to move the `RockerSwitchToPlayPauseProfile` and `RockerSwitchToRollershutterProfile` as they are not at all specific to EnOcean, but make sense for any SYSTEM_RAWROCKER in general.

> we have to listen to the physical rocker switches to sync the item state with the physical state.

We actually have a very similar situation in KNX, where we have a GA (group address) on which we send out commands and we have ["listening-GA"s](https://www.openhab.org/addons/bindings/knx/#group-address-notation), where we interpret the telegrams as state updates for items (while they are also "commands" within the KNX system, i.e. they make the addressee switch). In KNX, these addresses are coded as a single string (i.e. not as single parameters). Could that be an option for EnOcean as well? Do you need to provide more than the EnOcean Id of the physical rockers? With such a configuration, the "virtualRocker" channel implementation could then easily deal with the whole situation (and you wouldn't need ANY kind of profile for them).

fruggy83 commented:
@kaikreuzer sorry, I had problems with my private git server this weekend, so I used github to sync my work between my dev machines. I was in hurry and pushed to fast. The refactoring of the ClassicDevice is still WIP. I will hopefully finish the work within the next few days.
By the way thanks a lot for your hints regarding the KNX binding. I am using extensible channels for the listening part of the ClassicDevice now 👍

kaikreuzer commented:
Thanks @fruggy83, sounds good!
FYI: It would be cool to have that binding in the 2.4 release for which I'd plan a feature freeeze by end of this month - so not much time left for us here :-)

fruggy83 commented:
Hi @kaikreuzer I hopefully finished the work on the ClassicDevice thing. Took longer as expected as I was not happy with the implementation. I fixed up everything, removed the exports and updated the readme. I did not removed the PlayerProfile yet, as I first want to create a PR in ESH.

Best regards
Daniel

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/EnOceanProfileTypes.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class EnOceanProfileTypes {
+
+    public static final ProfileTypeUID RockerSwitchToPlayPause = new ProfileTypeUID(BINDING_ID,
```
As all those are constants, you should change the name to all upper case.

fruggy83 replied:
removed all profiles => fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromOnOffProfile.java:
```diff
@@ -0,0 +1,125 @@
+                    break;
+
+            }
+            if (switchMode == SwitchMode.RockerSwitch) {
```
What is this?

fruggy83 replied:
removed all profiles => fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromOnOffProfile.java:
```diff
@@ -0,0 +1,125 @@
+    @Override
+    public void onCommandFromHandler(Command command) {
+        // a rocker switch command should not be submitted to item
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromOnOffProfile.java:
```diff
@@ -0,0 +1,125 @@
+    @Override
+    public void onStateUpdateFromHandler(State state) {
+        // Rocker switches do not own a state
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromUpDownProfile.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Override
+    public void onCommandFromItem(Command command) {
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchToPlayPauseProfile.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        // nothing to do here => channel is readonly
+
```
remove empty line

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -108,6 +108,12 @@
+        <feature>esh-config-discovery-usbserial</feature>
+        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.enocean/${project.version}</bundle>
+    </feature>
     <feature name="openhab-binding-exec" description="Exec Binding" version="${project.version}">
```
please add an empty line

fruggy83 replied:
fixed

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -108,6 +108,12 @@
+    <feature name="openhab-binding-enocean" description="EnOcean Binding" version="${project.version}">
+        <feature>openhab-runtime-base</feature>
+        <feature>openhab-transport-serial</feature>
+        <feature>esh-config-discovery-usbserial</feature>
```
This is already included in `openhab-transport-serial`, so you can remove this line.

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromUpDownProfile.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void onCommandFromHandler(Command command) {
+        // a rocker switch command should not be submitted to item
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromUpDownProfile.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void onStateUpdateFromHandler(State state) {
+        // Rocker switches do not own a state
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromUpDownProfile.java:
```diff
@@ -0,0 +1,110 @@
+                            duration, TimeUnit.MILLISECONDS);
+                }
+            }
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromUpDownProfile.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        // do not react on item state updates as they are triggered by this profile
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/RockerSwitchFromOnOffProfile.java:
```diff
@@ -0,0 +1,125 @@
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        // do not react on item state updates as they are triggered by this profile
+
```
remove empty line

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanBindingConstants.java:
```diff
@@ -0,0 +1,295 @@
+    public final static String CHANNEL_GENERIC_COLOR = "genericColor";
+    public final static String CHANNEL_GENERIC_TEACHINCMD = "genericTeachInCMD";
+
+    public final static ChannelTypeUID VirtualRockerSwitchAChannelType = new ChannelTypeUID(BINDING_ID,
```
Please only use capitals for ALL constant names defined in this class.

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,300 @@
+
+## Thing Configuration
+
+The pairing process of an openHAB thing and an EnOcean device has to be triggered within PaperUI.
```
```suggestion
The pairing process of an openHAB thing and an EnOcean device has to be triggered within Paper UI.
```

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
```diff
@@ -0,0 +1,300 @@
+## Thing Configuration
+
+The pairing process of an openHAB thing and an EnOcean device has to be triggered within PaperUI.
+Therefore if you do not want to use PaperUI, a mixed mode configuration approach has to be done.
```
```suggestion
Therefore if you do not want to use Paper UI, a mixed mode configuration approach has to be done.
```

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/README.md:
````diff
@@ -0,0 +1,300 @@
+If you want to use a mapping transformation, your mapping file has to look like this for a genericThing using EEP A5_FF_FF:
+
+```
+ChannelId|OpenhabCommand=EnoceanData(Hex)
````
OpenhabCommand -> openHABCommand

fruggy83 replied:
fixed

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/profiles/EnOceanProfileTypes.java:
```diff
@@ -0,0 +1,50 @@
+    public static final ProfileTypeUID RockerSwitchToPlayPause = new ProfileTypeUID(BINDING_ID,
+            "rockerswitch-to-play-pause");
+
+    public static final ProfileTypeUID RockerSwitchFromOnOff = new ProfileTypeUID(BINDING_ID,
```
Do we actually still NEED the "rockerswitch-from..." profiles? I thought that the corresponding channel typ can already be used by the handler to determine the correct behavior. You aren't mentioning them in the README either.

fruggy83 replied:
I removed all profiles now. Logic of "rockerswitch-from..." profiles moved into the ClassicDeviceHandler where it belongs. I have updated the readme accordingly.

fruggy83 commented:
Hi @kaikreuzer, I fixed all comments and pushed an update. I will create the ESH issue/PR for the player profile tomorrow. However the current readme already mentions this profile (I hope this is ok).
So thank you very much for your patience and all your work.

Best regards
Daniel

kaikreuzer commented on addons/binding/org.openhab.binding.enocean/ESH-INF/config/config.xml:
```diff
@@ -41,5 +41,21 @@
+        <parameter name="duration" type="integer">
+                <label>Time in ms between Press and Release message</label>
+                <default>250</default>
+            </parameter>
```
Could you please reapply the formatter?

fruggy83 replied:
fixed

fruggy83 commented:
Hi @kaikreuzer, I fixed the formatting.
What should I do with the "Signed-off-by" validation? When I look at the details I see some commits missing the "Signed-off-by" but also commits with this text 🤔 So I am not sure, what is really missing here.
Thanks a lot for your quick review of the PlayerItem profile in ESH 👍

kaikreuzer commented:
> What should I do with the "Signed-off-by" validation?

No worries, I'll squash the commits on merge and keep the sign-off on the resulting one.

> Thanks a lot for your quick review of the PlayerItem profile in ESH

Just did a new ESH stable build, so it is directly available!

Hilbrand commented on addons/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_11/A5_11_04.java:
```diff
@@ -0,0 +1,196 @@
+                case WATT:
+                    factor = 1;
+                    break;
+                case KILOWATT:
```
This looks like a bug. `KILOWATT` falls through to `MEGAWATT`. I think it should be the other way around. Like on line 121 with `GIGAWATTHOUR`.

dominikkv replied:
Fixed this in #5054 as I am fixing another factor value there.
